### PR TITLE
Prevent checks from downloading whole repos

### PR DESCRIPTION
**Pull request by david-baylibre**
Repo versions are computed from manifests with updated revisions for each project. The repo tool can only generate a manifest with accurate sha1 hashes once all projects have been downloaded locally (repo manifest -r -o new-manifest.xml). This uses a lot of network bandwith, memory, cpu and disk space and is a huge waste of time.
git ls-remote is used instead to fetch revisions much quicker before being injected into the original manifest. A new variable check_jobs is used to spawn concurrent processes to make it close to X times faster
Adjust jobs and check_jobs variables based on the git servers capabilities/limitations

### CODE REVIEW

**Comment by makohoek** [comment]
The lint errors are still failing. See: 

https://github.com/makohoek/repo-resource/actions/runs/7788670056/job/21238599510?pr=28

File: repo_resource/test_check.py [hunk 1]
```diff
@@ -306,14 +315,14 @@ def test_ssh_private_key_without_manifest_access(self):
         with self.assertRaises(SystemExit):
             versions = check.check(instream)
 
-        self.assertEquals(len(versions), 0)
```
Comment (makohoek) [inline repo_resource/test_check.py:309]: These seem irrelvant to the commit and are solved in:

https://github.com/makohoek/repo-resource/pull/29

I will merge the above PR.

File: repo_resource/check.py [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 from repo_resource import common
 
+DEFAULT_CHECK_JOBS = 2
```
Comment (makohoek) [inline repo_resource/check.py:21]: There is no need for this. Instead, we should initialize:

```
    check_jobs: int = 2
```

In `class SourceConfiguration(NamedTuple)`

Having 2 variables for a single thing is error prone. What happens if we update this in the future?

File: repo_resource/check.py [hunk 3]
```diff
@@ -33,13 +35,23 @@ def check(instream) -> list:
...
         common.add_private_key_to_agent(config.private_key)
 
+    jobs = config.jobs
+    check_jobs = config.check_jobs or jobs*2 or DEFAULT_CHECK_JOBS
+
     try:
-        repo = common.Repo()
-        repo.init(config.url, config.revision, config.name, config.depth)
-        repo.sync(jobs=config.jobs)
+        repo = common.Repo(config.url,
```
Comment (makohoek) [inline repo_resource/check.py:49]: Why the formatting change?

Doing arbitrary format changes hurts the review as it makes the patch bigger (and thus more difficult to review)

If we want to keep the formatting change, should do it in a separate commit.

https://www.aleksandrhovhannisyan.com/blog/atomic-git-commits/

File: repo_resource/common.py [hunk 4]
```diff
@@ -12,17 +12,33 @@
...
 
 import ssh_agent_setup
 from repo import manifest_xml
 from repo import main as repo
 
+
 CACHEDIR = Path('/tmp/repo-resource-cache')
+SHA1_PATTERN = re.compile(r'^[0-9a-f]{40}$')
+EXCLUDE_ATTRS = {'dest-branch', 'upstream'}
+DEFAULT_CHECK_JOBS = 2
```
Comment (makohoek) [inline repo_resource/common.py:33]: Duplication with check.py

File: repo_resource/common.py [hunk 5]
```diff
@@ -56,6 +72,42 @@ def remove_private_key_from_agent():
...
+    without downloading the whole repo
+    """
+    # v1.0^{} is the commit referring to tag v1.0
+    # git ls-remote returns the tag sha1 if left as is
+    if branch.startswith('refs/tags'):
+        branch += '^{}'
+    try:
+        with redirect_stdout(sys.stderr):
+            # return tuple (remote/project, revision)
+            print('Fetching revision for '+remote+'/'+project+'...')
```
Comment (makohoek) [inline repo_resource/common.py:96]: Use modern printing with formatters please:

```
print('Fetching revision for {}/{} ...'.format(remote, project))
```

See how other prints are done in the project

File: repo_resource/common.py [hunk 6]
```diff
@@ -56,6 +72,42 @@ def remove_private_key_from_agent():
...
+            if is_sha1(branch):
+                return (remote + '/' + project, branch)
```
Comment (makohoek) [inline repo_resource/common.py:98]: same here: `return '{}/{}'.format(project, branch)`

File: repo_resource/common.py [hunk 7]
```diff
@@ -56,6 +72,42 @@ def remove_private_key_from_agent():
...
+            g = git.cmd.Git()
+            url, revision = (
+                remote + '/' + project,
+                g.ls_remote(remoteUrl+'/'+project, branch).split()[0]
+            )
+            print(url+': '+revision)
+            return (url, revision)
+    except Exception as e:
+        print('Cannot fetch project ', remoteUrl+'/'+project)
```
Comment (makohoek) [inline repo_resource/common.py:107]: This won't be shown in stderr so concourse will not display it

What's the purpose of these error messages in that case?

Should we print this to stderr instead?

File: repo_resource/common.py [hunk 8]
```diff
@@ -76,8 +129,10 @@ def source_config_from_payload(payload):
     p = SourceConfiguration(**payload['source'])
     source_url = urlparse(p.url)
 
-    if source_url.netloc == 'gitlab.com' and \
-       (source_url.scheme == 'http' or source_url.scheme == 'https'):
+    if (
+        source_url.netloc == 'gitlab.com' and
+        re.fullmatch('https?', source_url.scheme)
```
Comment (makohoek) [inline repo_resource/common.py:134]: Seems like a good improvement, but it's not relevant to this patch.

This patch is "use git ls-remote for fetching revision"

This diff is "improve url matching regex".

Please split this out to a seperate commit.

File: repo_resource/common.py [hunk 9]
```diff
@@ -105,6 +160,28 @@ def to_file(self, filename):
     def metadata(self) -> str:
         return ''
 
+    def standard(self) -> str:
+        try:
+            root = ET.fromstring(self.__version)
+            for element in root:
+                if element.tag not in TAGS:
+                    root.remove(element)
+                    print("Removed ", element.tag)
```
Comment (makohoek) [inline repo_resource/common.py:169]: Please don't print things. Printing things to stdout will break concourse checking mechanism

File: repo_resource/common.py [hunk 10]
```diff
@@ -105,6 +160,28 @@ def to_file(self, filename):
...
+            sorted_xml = sorted(root, key=lambda x: (
+                TAGS.index(x.tag) if x.tag in TAGS else 999,
```
Comment (makohoek) [inline repo_resource/common.py:171]: 999 ? what's this number for?

File: repo_resource/common.py [hunk 11]
```diff
@@ -121,9 +198,16 @@ class Repo:
     such as init/sync and manifest
     """
 
-    def __init__(self, workdir=CACHEDIR):
+    def __init__(self, url, revision='HEAD', name='default.xml',
+                 depth=-1, workdir=CACHEDIR):
         self.__workdir = workdir
         self.__oldpwd = None
+        self.__url = url
```
Comment (makohoek) [inline repo_resource/common.py:205]: adding these parameters could probably be a seperate (preparatory) patch

File: repo_resource/common.py [hunk 12]
```diff
@@ -185,34 +277,33 @@ def sync(self, version: Version = None, jobs: int = 0):
...
-                            '--manifest-name={}'.format(tmp_manifest))
-                        repo._Main(repo_cmd)
+                if os.listdir(self.__workdir) == []:
+                    raise Exception('Sync failed. Is manifest correct?')
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    tmp_manifest = os.path.join(tmpdir, 'manifest_snapshot')
+                    self.__manifest_out(tmp_manifest)
+                    self.__version = Version.from_file(tmp_manifest)
         except Exception as e:
+            sys.exit(1)
```
Comment (makohoek) [inline repo_resource/common.py:293]: Why is this added? shouldn't we restore the old pwd if an exception happens ?

File: repo_resource/requirements.txt [hunk 13]
```diff
@@ -1,2 +1,3 @@
-gitrepo==2.31.1
+gitrepo==2.32.2
```
Comment (makohoek) [inline repo_resource/requirements.txt:1]: why are these updates needed? I think this could be done in a seperate patch

File: repo_resource/test_in.py [hunk 14]
```diff
@@ -4,22 +4,22 @@
...
 
 from . import check
 from . import common
 from . import in_
 
 
 class TestIn(unittest.TestCase):
-
     def setUp(self):
+        os.chdir('/root')
```
Comment (makohoek) [inline repo_resource/test_in.py:22]: why ?

**Review by makohoek — CHANGES_REQUESTED** [review]
Hi @david-baylibre ,

Thank you for this PR. It's a really nice improvement.

I had some difficulties to review this in detail because it's a single big commit doing many things.

For the next version, could you please consider splitting this ?

Here is an example of how this could be split:

1. requirements: bump up gitrepo version
2. common: improve url matching regex for http/https
3. common: add url,revision,name,depth when creating a Repo
4. common: formatting changes (if you wish to have these)
5. Implement getRevision using git ls-remote (and add unit tests for them)
6. Migrate CHECK to use git-ls-remote

Does that make sense ?

File: repo_resource/check.py (hunk 3 above)
Comment (david-baylibre) [reply repo_resource/check.py:49]: More arguments so the line was too long to be PEP8 compliant

File: repo_resource/common.py (hunk 4 above)
Comment (david-baylibre) [reply repo_resource/common.py]: Using the variable in common everywhere now

File: repo_resource/common.py (hunk 5 above)
Comment (david-baylibre) [reply repo_resource/common.py]: Done

File: repo_resource/common.py (hunk 7 above)
Comment (david-baylibre) [reply repo_resource/common.py]: Printed in stderr now

File: repo_resource/common.py (hunk 8 above)
Comment (david-baylibre) [reply repo_resource/common.py:143]: Done

File: repo_resource/common.py (hunk 10 above)
Comment (david-baylibre) [reply repo_resource/common.py:182]: 999 is the default position for unknown elements in the manifest
Comments added in code with more details

File: repo_resource/common.py (hunk 11 above)
Comment (david-baylibre) [reply repo_resource/common.py:216]: Done

File: repo_resource/common.py (hunk 12 above)
Comment (david-baylibre) [reply repo_resource/common.py]: Removed. Tests have been changed to match expected errors

File: repo_resource/requirements.txt (hunk 13 above)
Comment (david-baylibre) [reply repo_resource/requirements.txt:1]: Keeping dependencies up to date
Created a separate commit

File: repo_resource/test_in.py (hunk 14 above)
Comment (david-baylibre) [reply repo_resource/test_in.py]: Most methods start moving to the working directory and move back to the previous folder, just like cd /dir; cd -
The issue with this is we end up in the working directory forever if a method calls another method eg
cd /dir; (cd /dir; cd -); cd -, we stay in /dir...
I skipped this issue calling methods directly and removed this addition.
It'd be good to change this to some simple cd /dir; cd $HOME which would make things easier in my opinion

**Review by makohoek — CHANGES_REQUESTED** [review]
Hi David,

Thank you for this new version.

I'm seeing some regression on URLS where we use ssh instead of https (with `make tests`)
https://gist.github.com/makohoek/c13a71ea7513403c42d64b4ac983fc3d

I would like to see those solved before merging.

Anyways, this is way better than the previous one.

I'm not 100% satisfied with the current design. My main complaint is that the `Version` class now has 2 representation of the version (a normalized one and the repo version string.)

I also would have like to see some update in the README documentation.
Right now, we don't descibe the new behaviour is not documented in the README:
https://github.com/makohoek/repo-resource?tab=readme-ov-file#check-check-for-new-versions-in-each-project-under-a-manifest

Both are no deal-breakers so I'm willing to merge this.
We can create github issues to follow up on those.

**Comment by makohoek** [comment]
Created:
- https://github.com/makohoek/repo-resource/issues/31
- https://github.com/makohoek/repo-resource/issues/32 

to follow up.

**Review by makohoek — APPROVED** [review]
